### PR TITLE
Fix multiple results not returned with custom TestMethod and TestClass (#358)

### PR DESCRIPTION
**Pull request by bignoncedric**
This PR fixes the bug #358.
In `TestMethodRunner`, only the first element of `TestResult[]` returned by `Executor.Execute` was used.

A test is also added to validate the [RFC 003- Framework Extensibility for Custom Test Execution](https://github.com/Microsoft/testfx-docs/blob/master/RFCs/003-Customize-Running-Tests.md).

### CODE REVIEW

File: src/Adapter/MSTest.CoreAdapter/Execution/TestMethodRunner.cs [hunk 1]
```diff
@@ -247,22 +247,28 @@ internal UnitTestResult[] RunTestMethod()
...
+                                            new UTF.TestResult() { TestFailureException = new Exception(string.Format(CultureInfo.CurrentCulture, Resource.UTA_ExecuteThrewException, ex.Message), ex) }
+                                        };
                                     }
 
-                                    currentResult.DatarowIndex = rowIndex++;
                                     watch.Stop();
-                                    currentResult.Duration = watch.Elapsed;
+                                    foreach (var testResult in testResults)
+                                    {
+                                        testResult.DatarowIndex = rowIndex++;
```
Comment (jayaranigarg) [inline src/Adapter/MSTest.CoreAdapter/Execution/TestMethodRunner.cs]: DatarowIndex is used for each row of data in dataSource. 
By framework extensibility you have now started running each dataRow multiple times(say 5), but we should not increase dataRowIndex for each of the 5 runs of a DataRow. It should be incremented only once per dataRow and not once per testResult.
This statement should be testResult.DataRowIndex = rowIndex;
Do rowIndex++ separately outside of the loop.

File: test/E2ETests/Automation.CLI/CLITestBase.cs [hunk 2]
```diff
@@ -160,6 +160,7 @@ public void ValidatePassedTestsContain(params string[] passedTests)
         {
             foreach (var test in passedTests)
             {
+                var x = string.Join(", ", this.runEventsHandler.PassedTests.Select(_ => _.DisplayName));
```
Comment (jayaranigarg) [inline test/E2ETests/Automation.CLI/CLITestBase.cs]: Remove this.

File: test/E2ETests/TestAssets/FxExtensibilityTestProject/CustomTestExTests.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+        private static int customTestClass1ExecutionCount;
+        [TestMethod]
+        public void CustomTestClass1()
+        {
+            customTestClass1ExecutionCount++;
+            Assert.AreNotEqual(3, customTestClass1ExecutionCount);
+        }
+    }
+
```
Comment (jayaranigarg) [inline test/E2ETests/TestAssets/FxExtensibilityTestProject/CustomTestExTests.cs:34]: Consider adding a data-driven test with [IterativeTestMethod] attribute

**Comment by jayaranigarg** [comment]
Kindly sign the CLA agreement as well.

**Comment by jayaranigarg** [comment]
Thank you for your contribution @bignoncedric . This PR looks awesome. 
As per our offline conversation, I am keeping this PR on hold. Let me know whenever you are ready to push this in.

**Comment by jayaranigarg** [comment]
@bignoncedric : Any updates here?

**Comment by federicoce-moravia** [comment]
Any updates? I'm currently sorting the `ITestResult[]` with a "worst outcome first" comparer before returning, but I would really like to see this handled elegantly by the library.

**Comment by ShreyasRmsft** [comment]
@bignoncedric any updates?

**Comment by jayaranigarg** [comment]
@dotnet-bot Test Windows / Debug Build please

**Pull request merged**